src/java.desktop/share/classes/com/sun/java/swing/SwingUtilities3.java

```diff
@@ -135,4 +143,64 @@ public static RepaintManager getDelegateRepaintManager(Component
         }
         return delegate;
     }
+
+    @FunctionalInterface
+    public interface UnscaledBorderPainter {
+        void paintUnscaledBorder(Component c, Graphics g, int x, int y,
+                                 int w, int h, double scaleFactor);
+    }
+
+    public static void paintBorder(Component c, Graphics g, int x, int y,
+                            int w, int h, UnscaledBorderPainter painter) {
+
+        // Step 1: Reset Transform
+        AffineTransform at = null;
+        Stroke oldStk = null;
+        boolean resetTransform = false;
+        double scaleFactor = 1;
+
+        int xtranslation = x;
+        int ytranslation = y;
+        int width = w;
+        int height = h;
+
+        if (g instanceof Graphics2D) {
+            Graphics2D g2d = (Graphics2D) g;
+            at = g2d.getTransform();
+            oldStk = g2d.getStroke();
+            scaleFactor = Math.min(at.getScaleX(), at.getScaleY());
+
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // or if scale=1, skip resetting the transform
+            resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0)) &&
+                    ((at.getScaleX() > 1) || (at.getScaleY() > 1));
+
+            if (resetTransform) {
+                /* Deactivate the HiDPI scaling transform,
+                 * so we can do paint operations in the device
+                 * pixel coordinate system instead of the logical coordinate system.
+                 */
+                g2d.setTransform(new AffineTransform());
+                double xx = at.getScaleX() * x + at.getTranslateX();
+                double yy = at.getScaleY() * y + at.getTranslateY();
+                xtranslation = clipRound(xx);
+                ytranslation = clipRound(yy);
+                width = clipRound(at.getScaleX() * w + xx) - xtranslation;
+                height = clipRound(at.getScaleY() * h + yy) - ytranslation;
+            }
+        }
+
+        g.translate(xtranslation, ytranslation);
+
+        // Step 2: Call respective paintBorder with transformed values
+        painter.paintUnscaledBorder(c, g, x, y, width, height, scaleFactor);
+
+        // Step 3: Restore Transform
+        g.translate(-xtranslation, -ytranslation);
+        if (resetTransform) {
+            Graphics2D g2d = (Graphics2D) g;
+            g2d.setTransform(at);
+            g2d.setStroke(oldStk);
```

Comment:
The stroke should always be restored even if the transform wasn't reset. Or it should be done in the implementation of `paintUnscaledBorder`.

Reply:
@aivanov-jdk Earlier strokeWidth was set when resetTransform was true and it was restored on the same condition. Now since the stroke width is moved to individual implementation of paintUnscaledBorder, it make sense to restore it outside of resetTranform.

To avoid missing this step, I have restored it in common code than in individual paintUnscaledBorder implementations.